lua/sg/auth.lua

```diff
@@ -0,0 +1,137 @@
+---@tag sg.auth
+---@config { ["module"] = "sg.auth" }
+
+---@brief [[
+--- How to manage authentication for Sourcegraph within Neovim
+--- (both for Cody and for Sourcegraph)
+---
+---@brief ]]
+
+local config = require "sg.config"
+local data = require "sg.private.data"
+
+local json_or_nil = require("sg.utils").json_or_nil
+local strategy = require("sg.types").auth_strategy
+
+local M = {}
+
+local valid = function(s)
+  return s and type(s) == "string" and s ~= ""
+end
+
+-- TODO: It does feel a bit weird to read the files all the time,
+-- but I think it's alright for now -- it also makes the state
+-- always correct with whatever their current file is.
+
+--- The default strategies for sg.nvim. Use |config.auth_strategy| to configure the
+--- order of evaluation. Whichever one returns a valid configuration first will be used
+--- when starting and connecting to Sourcegraph and Cody.
+---@eval { ["description"] = require('sg.auth').__docs() }
+---@type table<SourcegraphAuthStrategy, SourcegraphAuthObject>
+M.strategies = {
+  [strategy.app] = {
+    doc = "Use the Cody App configuration to connect to your sourcegraph instance."
+      .. " See https://sourcegraph.com/get-cody for more information",
+    get = function()
+      local locations = {
+        "~/Library/Application Support/com.sourcegraph.cody/app.json",
+      }
+
+      if vim.env.XDG_DATA_HOME then
+        table.insert(locations, vim.env.XDG_DATA_HOME .. "/com.sourcegraph.cody/app.json")
+      end
+
+      table.insert(locations, vim.fn.expand "$HOME/.local/share/com.sourcegraph.cody/app.json")
+
+      -- TODO: Handle windows paths
+      -- table.insert(vim"{FOLDERID_LocalAppData}/com.sourcegraph.cody/app.json",
+
+      for _, file in ipairs(locations) do
+        local parsed = json_or_nil(file)
+        if parsed and valid(parsed.token) and valid(parsed.endpoint) then
+          return { token = parsed.token, endpoint = parsed.endpoint }
+        end
+      end
+
+      return nil
+    end,
```

Comment:
This makes me wonder whether it shouldn't be possible to get the tokens directly from the Cody App itself? Rather than reading from where the Cody App ends up storing the tokens.

Not a blocker, but does feel weird to "peek behind the curtains" for auth

Reply:
Yeah, that's just the current way App is doing it and the other editors. Agreed would not mind a diff strat at some point though.